File: 22-generate-parentheses/generate-parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> out;
        int l=0,r=0;
        f(out, "",l,r,n);
        return out;
    }
    void f(vector<string>&out, string s,int l, int r,int n){
        if(l+r==n*2) out.push_back(s);
        else{
            if(l<n) f(out,s+"(",l+1,r,n);
            if(r<l) f(out,s+")",l,r+1,n);
        }
        return;
    }

};

File: 22-generate-parentheses/generate-parentheses.java
class Solution {
    public void helper(List<String> ans, int openCount, int closeCount, String str, int n) {
        // Base Case
        if (openCount == n && closeCount == n) {
            ans.add(str);
        } else {
            // Sub Problem
            if (openCount > closeCount) {
                helper(ans, openCount, closeCount + 1, str + ")", n);
            }
            if (openCount < n) {
                helper(ans, openCount + 1, closeCount, str + "(", n);
            }
        }
    }

    public List<String> generateParenthesis(int n) {
        List<String> ans = new ArrayList<>();
        int openCount = 0;
        int closeCount = 0;
        String str = "";
        helper(ans, openCount, closeCount, str, n);
        return ans;
    }
}